Cms export validation scaffolding

cvalarida opened the pull request:
## Description
Validates the content model using [JSON schema](https://www.npmjs.com/package/jsonschema).

## Testing done
Manual.

## Screenshots
![image](https://user-images.githubusercontent.com/12970166/67995550-2ad43900-fc08-11e9-82ff-bcc59ee952b4.png)

## Acceptance criteria
- [x] Content is validated according to an explicit schema

Commit:
Update comment

Commit:
Validate content models

kelsonic commented on src/site/stages/build/page-builder/schemas/page.js:
```diff
@@ -0,0 +1,51 @@
+/* eslint-disable camelcase */
```
Interesting! What is a `page` entity? I don't think I included it in [the docs](https://github.com/department-of-veterans-affairs/vets-website/tree/kelsonic-2791-tome-sync-testing-scaffolding/src/site/stages/build/page-builder/tests/entities).

cvalarida replied:
It's a type of `node`.
```
❯ rg '"target_id": "page"' | wc -l
221
```

kelsonic commented on src/site/stages/build/tome-sync-page-builder.js:
```diff
@@ -33,10 +35,20 @@ const assembleEntityTree = (entityType, uuid, parents = []) => {
   }
 
-  const entity = getModifiedEntity(entityType, readEntity(entityType, uuid));
+  const rawEntity = readEntity(entityType, uuid);
```
Hmm, we're touching a lot of the same files and loc haha

ncksllvn commented on src/site/stages/build/page-builder/schemas/#page.js#:
```diff
@@ -0,0 +1,51 @@
+/* eslint-disable camelcase */
```
We're definitely gonna have to convert all of this to camelcase at some point. It might make sense to do right out of the gate, so that the schemas align with the data structures in the templates. What do you think?

cvalarida replied:
That'll happen in the transformation step after this. The goal of this validation is to ensure that the data we get in the raw is what we expect.

ncksllvn commented on src/site/stages/build/page-builder/validator.js:
```diff
@@ -0,0 +1,20 @@
+
+// Set up formats
+Validator.prototype.customFormats.uuid = input =>
+  /[a-z0-9]{8}-[a-z0-9]{4}-[a-z0-9]{4}-[a-z0-9]{4}-[a-z0-9]{12}/.test(input);
```
Do we need to validate the UUID? Couldn't we just accept it as a string?

cvalarida replied:
Right, I was honestly kind of stretching here. Mostly this serves as an example of adding a custom format in case we need it later. I'm totally fine with moving it.

cvalarida replied:
Turns out when using `ajv`, because it supports the latest JSON schema draft (07), this isn't needed anymore; `uuid` is a built-in format.

ncksllvn commented on src/site/stages/build/page-builder/schema-validation.js:
```diff
@@ -0,0 +1,39 @@
+    if (!missingSchemas.has(contentModelType)) {
+      missingSchemas.add(contentModelType);
+      // eslint-disable-next-line no-console
+      console.warn(`Missing schema for ${contentModelType}`);
```
All of these console statements are pretty useful, and maybe we should encourage that. What do you think of using something a module like `debug` so that we can see all of this output while non-CMS engineers don't need to be bothered by it? I think it would be started with something like - `DEBUG=cms npm run build-content`

ncksllvn commented on src/site/stages/build/page-builder/schema-validation.js:
```diff
@@ -0,0 +1,39 @@
+
+const schemas = {
+  page,
+};
```
There will likely be a lot of schemas, so it might be useful to use a helper function that does a dynamic `require` under-the-hood to import the schema.

ncksllvn commented on src/site/stages/build/page-builder/schemas/common/entity-reference.js:
```diff
@@ -0,0 +1,23 @@
+};
+
+const entityReferenceArray = {
+  id: '/EntityReferenceArray',
```
Just curious (pardon how little I know about JSON schema) - how come these need leading slashes?

cvalarida replied:
No slashes needed. Removed!

ncksllvn reviewed (commented):
Love this!

ncksllvn commented:
With that empty array issue @cvalarida , could this be relevant? https://github.com/ruby-json-schema/json-schema/issues/311#issuecomment-189233598

Commit:
Rename ids

Commit:
Require field_alert and field_page_last_built

Commit:
Remove unused ids

Commit:
Ignore alerts for now

Turns out some of these are empty arrays, but some are arrays of a
single empty array.

```js
"field_alert": [ [] ]
```

Because why not?

Commit:
Use ajv for json schema validation

Commit:
Abort!

cvalarida commented:
Turns out some of the `field_alert`s were `[ [] ]` :roll_eyes: 

So I'm just ignoring those for now. Until we know what their shape _should_ be (if we use them at all, that is).

bkjohnson commented on src/site/stages/build/page-builder/schemas/#page.js#:
```diff
@@ -0,0 +1,51 @@
+          value: { type: 'string' },
+        },
+      },
+    },
```
Is there a reason why this isn't a `{ $ref: 'generic-nested-string' }`?

cvalarida replied:
Blast. Because it was a temp file that shouldn't have gotten committed. :shakes-fist:

bkjohnson replied:
I didn't even notice the filename!

bkjohnson commented on src/site/stages/build/page-builder/schemas/page.js:
```diff
@@ -0,0 +1,51 @@
+      items: {
+        type: 'object',
+        properties: {
+          value: { type: 'string', format: 'custom-date-time' },
```
For situations like this would it be useful to have some helper that takes a `format` argument that it uses to modify a `genericNestedString` with?

cvalarida replied:
Yeah, we could do that if we run into this a lot. Alternatively, what we could do is create a new schema for the nested custom-date-time.

bkjohnson commented on src/site/stages/build/page-builder/schemas/page.js:
```diff
@@ -0,0 +1,51 @@
+    'field_related_links',
+    'field_administration',
+    'field_page_last_built',
+  ],
```
If this is supposed to match what we are extracting in the filtering step I can see it quickly getting out of sync.  Should we have a single source of truth for our required/filtered items?

cvalarida replied:
Good call! Yes. I'll pull in the filters. :+1:

Commit:
Merge branch 'master' into cms-export-validation-scaffolding-cv

Commit:
Add baseType and uuid in readEntity; stop passing around entityType

Commit:
Delete temp file

Commit:
Use getFilter in page schema

bkjohnson commented on src/site/stages/build/process-cms-exports/schemas/page.js:
```diff
@@ -0,0 +1,42 @@
+
+module.exports = {
+  type: 'object',
+  properties: {
```
Should we also have `metatag` in here now?

cvalarida replied:
Probably. I need to download the one with metatags...

cvalarida replied:
[Done](https://github.com/department-of-veterans-affairs/vets-website/pull/10982/commits/eaccbba69ed7107020544a4ab601d970b43baeb8)!

Commit:
Add metatag property to page schema

Commit:
Dynamically import all the schemas

bkjohnson reviewed (approved):
:+1:

Commit:
Merge branch 'master' into cms-export-validation-scaffolding-cv

Commit:
Fix failure after merge; filter and transform in assembleEntityTree
